### PR TITLE
#2009

### PR DESCRIPTION
**Pull request by jshaw86**
Per Woody's comments, i've made the following changes:
1) `or_join_open()` and friends changed to `or_on_open()`
2) added `or_on()` method, was supported through `->on('col','=','col','OR')` but i agree this is cleaner
3) removed .gitignore(my bad)
4) rebased and merged 3.1/develop
5) In terms of supporting everything in #2009, this should be close, but this was implemented before I saw #2009(to suffice a need at work) so if there are some flagrant inconsistencies let me know and I would be glad to make things align more #2009
I've tested the following and copied the results:
    echo Kohana_DB::select()->from(array('foo','bar'))
         ->join(array('baz','bam'))
         ->on('bar.col3','=','bam.col3')
        ->or_on_open()
          ->on('bar.col','=','bam.col')
          ->or_on('bar.col2','=','bam.col2')
        ->or_on_close();

```
 SELECT * FROM `foo` AS `bar` JOIN `baz` AS `bam` ON `bar`.`col3` = 'bam.col3' OR (`bar`.`col` = 'bam.col' OR `bar`.`col2` = 'bam.col2')
```
## 

```
echo Kohana_DB::select()->from(array('foo','bar'))
    ->join(array('baz','bam'))
    ->on('bar.col3','=','bam.col3')
    ->and_on_open()
      ->on('bar.col','=','bam.col')
      ->or_on('bar.col2','=','bam.col2')
    ->or_on_close();

 SELECT * FROM `foo` AS `bar` JOIN `baz` AS `bam` ON `bar`.`col3` = 'bam.col3' AND (`bar`.`col` = 'bam.col' OR `bar`.`col2` = 'bam.col2')
```
## 

```
  echo Kohana_DB::select()->from(array('foo','bar'))
    ->join(array('baz','bam'))
    ->and_on_open()
      ->on('bar.col','=','bam.col')
      ->or_on('bar.col2','=','bam.col2')
    ->or_on_close()
    ->on('bar.col3','=','bam.col3');

SELECT * FROM `foo` AS `bar` JOIN `baz` AS `bam` ON (`bar`.`col` = 'bam.col' OR `bar`.`col2` = 'bam.col2') AND `bar`.`col3` = 'bam.col3'
```
## 

```
  echo Kohana_DB::select()->from(array('foo','bar'))
    ->join(array('baz','bam'))
    ->or_on_open()
      ->on('bar.col','=','bam.col')
      ->or_on('bar.col2','=','bam.col2')
    ->or_on_close()
    ->or_on('bar.col3','=','bam.col3');

SELECT * FROM `foo` AS `bar` JOIN `baz` AS `bam` ON (`bar`.`col` = 'bam.col' OR `bar`.`col2` = 'bam.col2') OR `bar`.`col3` = 'bam.col3'
```

### CODE REVIEW

File: classes/kohana/database/query/builder/join.php [hunk 1]
```diff
@@ -99,39 +146,40 @@ class Kohana_Database_Query_Builder_Join extends Database_Query_Builder {
 		{
 			$sql = 'JOIN';
 		}
+		if(is_array($this->_table) && ($this->_table[0] instanceof Kohana_Database_Query)){
+			// Quote the table name that is being joined
+			$sql .= ' ( '.$this->_table[0].' ) AS `'.$this->_table[1].'` ON ';
```
Comment (shadowhand) [inline classes/kohana/database/query/builder/join.php]: You don't need to do this, it will be automatically handled by `Database::quote_table`.

File: classes/kohana/database/query/builder/join.php [hunk 2]
```diff
@@ -99,39 +146,40 @@ class Kohana_Database_Query_Builder_Join extends Database_Query_Builder {
...
+		}
+		else{
+			// Quote the table name that is being joined
+			$sql .= ' '.$db->quote_table($this->_table).' ON ';
+		}
+		
+		$sql .= $this->_compile_conditions($db, $this->_on);
```
Comment (shadowhand) [inline classes/kohana/database/query/builder/join.php]: Using `_compile_conditions` will not work here, it will assume the right hand (`left operator right`) part of the SQL fragment is a value rather than a column.

Comment (jshaw86) [inline classes/kohana/database/query/builder/join.php]: OK, that didn't get merged for some reason. I can fix this one of 2 ways.
1) and a preg_match to compile conditions where if the right is of the format `[a-zA-Z0-9]\.[a-zA-Z0-9]` then db->quote it, the down side to this approach is it requires the user to prefix columns with table names or alias's(probably a good practice in general, but isn't required by standard SQL).
2) override _compile_conditions in JOIN to db->quote the right hand

Do you have a preference?

Comment (shadowhand) [inline classes/kohana/database/query/builder/join.php]: I greatly prefer the 2nd option, even though is has a lot of duplication.

Comment (jshaw86) [inline classes/kohana/database/query/builder/join.php]: fixed here: https://github.com/jshaw86/database/commit/88a413ac2a429626311e939f655a5f90d94292cc, updated pull request commit range.

Comment (shadowhand) [inline classes/kohana/database/query/builder/join.php]: My last request is to rebase these commits into a single change so that the log history looks cleaner. Everything looks good now!

Comment (jshaw86) [inline classes/kohana/database/query/builder/join.php]: OK I'll try I've had limited success with git squashing. I'm going to get the tests in there before I do that though.

Comment (shadowhand) [inline classes/kohana/database/query/builder/join.php]: If you can't make it work, I'll just merge your changes with --squash and run them into a single commit. Unfortunately, this destroys history in a worse way than having you rebase.

**Comment by jshaw86** [comment]
OK I got it down to 3 commits, that's the best i could do. for some reason it dupped the "overrid Kohana_Database_Query_Builder ....."  commit i'm not sure how to fix that.

**Comment by jshaw86** [comment]
thought I would give a quick update on unit test progress: https://github.com/jshaw86/database/tree/db_tests

I have implemented tests for:
1. just inputting select columns
2. inputting select columns and from tables
3. inputting select columns, from tables and where conditions
4. started on joins but need to tweak a few things, probably finish it out tomorrow

I probably need to be adding tests for edge cases also, checking exception handling and stuff, that maybe tomorrow or next week sometime, any advice or criticism on what I have so far is welcome.
